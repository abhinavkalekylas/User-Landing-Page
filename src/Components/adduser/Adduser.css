.container {
    margin-top: 1rem;
}

.form, label {
    color: black;
}

.input {
    display: flex;
    flex-direction: row;
}


label {
    font-size: 1.2rem;
    display: flex;
    justify-content: center;
    align-items: center;
    margin: 1rem;
}

.btn_add {
    height: 3rem;
    margin: 1rem 11rem;   
}