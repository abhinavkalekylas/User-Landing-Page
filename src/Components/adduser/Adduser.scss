$text-color: black; 

.adduser_container {
    margin-top: 1rem;

    .form, .adduser_label {
        color: $text-color;
    }

    .input {
        display: flex;
        flex-direction: row;
    }

    .adduser_label {
        font-size: 1.2rem;
        display: flex;
        justify-content: center;
        align-items: center;
        margin: 0.5rem;
    }

    .btn_add, .reset_add {
        height: 3rem;
        margin: 1rem 4rem;   
    }
}
