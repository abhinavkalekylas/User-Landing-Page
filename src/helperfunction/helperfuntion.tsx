export interface User {
  id: number;
  name: string;
  email: string;
  gender: GenderType;
  status: StatusType;
}

export enum GenderType {
  MALE = "male",
  FEMALE = "female",
}

export enum StatusType {
  ACTIVE = "active",
  INACTIVE = "inactive",
}
