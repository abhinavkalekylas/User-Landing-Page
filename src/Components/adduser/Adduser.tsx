import React, { useEffect, useState } from "react";
import { useNavigate } from "react-router-dom";
import { Form, FormGroup, Button } from "reactstrap";
import axios from "axios";
import "./Adduser.css";
import {
  User,
  GenderType,
  StatusType,
} from "../../helperfunction/helperfuntion";

const AddUser = ({ handleClose }: { handleClose: () => void }) => {
  const navigate = useNavigate();
  const userData: User = {
    id: 0,
    name: "",
    email: "",
    gender: GenderType.MALE,
    status: StatusType.ACTIVE,
  };

  // form for user data after click add user button
  const [userForm, setUserForm] = useState(userData);
  const [error, setError] = useState(false);

  const handleChange = (e: { target: { name: any; value: any } }) => {
    setUserForm({
      ...userForm,
      [e.target.name]: e.target.value,
    });
  };

  const handleForm = async (e: { preventDefault: () => void }) => {
    e.preventDefault();
    if (userForm.name === "" || userForm.email === "") {
      return alert("Please enter required field");
    } else if (!validateEmail(userForm.email)) {
      return alert("Email is not valid");
    } else {
      try {
        const token =
          "3f30438c7b3212b121ae63e52bae216ca2bc11b700c8aa29cb0891d61cc96fca";
        const res = await axios("https://gorest.co.in/public/v2/users", {
          method: "POST",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Bearer ${token}`,
          },
          data: JSON.stringify(userForm),
        });
        navigate(0);
      } catch (error) {
        setError(true);
        console.log("Error while creating user ", error);
      }
    }
  };

  const validateEmail = (email: string) => {
    const regex = /^[^\s@]+@[^\s@]+\.[^\s@]{2,}$/i;
    if (regex.test(email)) {
      return true;
    }
    return false;
  };

  const resetForm = () => {
    setUserForm(userData);
  };

  return (
    <div className="container">
      {error ? (
        <h2>This page is under develop. We will sure give you an update</h2>
      ) : (
        <>
          <Form onSubmit={handleForm} className="form">
            <FormGroup>
              <div className="input input-group input-group-lg">
                <label>Name :</label>
                <input
                  className="username form-control my-2"
                  placeholder="Enter username here"
                  aria-label="Sizing example input"
                  aria-describedby="inputGroup-sizing-lg"
                  onChange={handleChange}
                  type="text"
                  name="name"
                  value={userForm.name}
                  autoComplete="off"
                  required
                />
              </div>
              <div className="input input-group input-group-lg">
                <label>Email :</label>
                <input
                  className="email form-control my-2"
                  placeholder="Enter email id here"
                  aria-label="Sizing example input"
                  aria-describedby="inputGroup-sizing-lg"
                  onChange={handleChange}
                  type="email"
                  name="email"
                  value={userForm.email}
                  autoComplete="off"
                  required
                />
              </div>
              <div className="input input-group input-group-lg">
                <label>Gender :</label>
                <select
                  name="gender"
                  value={userForm.gender}
                  onChange={handleChange}
                  className="gender form-select form-select-lg mb-3"
                  aria-label=".form-select-lg example"
                >
                  <option value={GenderType.MALE}>Male</option>
                  <option value={GenderType.FEMALE}>Female</option>
                </select>
              </div>

              <div className="input input-group input-group-lg">
                <label>Status :</label>
                <select
                  name="status"
                  value={userForm.status}
                  onChange={handleChange}
                  className="status form-select form-select-lg mb-3"
                  aria-label=".form-select-lg example"
                >
                  <option value={StatusType.ACTIVE}>Active</option>
                  <option value={StatusType.INACTIVE}>Inactive</option>
                </select>
              </div>
              <div className="d-flex">
                <Button
                  className="btn_add"
                  color="success"
                  type="submit"
                  onClick={handleClose}
                >
                  Add User
                </Button>

                <Button
                  className="reset_add"
                  color="primary"
                  onClick={resetForm}
                >
                  Reset
                </Button>
              </div>
            </FormGroup>
          </Form>
        </>
      )}
    </div>
  );
};

export default AddUser;
