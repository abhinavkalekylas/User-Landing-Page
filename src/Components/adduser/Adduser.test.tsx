import { BrowserRouter as Router } from "react-router-dom";
import { act, render, screen } from "@testing-library/react";
import axios from "axios";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import { GenderType, StatusType } from "../../helperfunction/helperfuntion";
import Adduser from "./Adduser";
import userEvent from "@testing-library/user-event";

let wrapper: any;
beforeEach(() => {
  const props = {
    handleClose: jest.fn(),
  };
  wrapper = shallow(
    <Router>
      <Adduser {...props} />
    </Router>
  );
});

// snapshot testing
it("renders correctly Adduser component", () => {
  expect(toJson(wrapper)).toMatchSnapshot();
});

// test input tag for name
it("test username input", () => {
  const event = { target: { value: "shivam" } };
  wrapper.find(".username").simulate("change", event);
  wrapper.update();
  expect(wrapper.find(".username").props().value).toEqual("shivam");
});

// test input tag for email
it("test email input", () => {
  const event = { target: { value: "shivam@gmail.com" } };
  wrapper.find(".email").simulate("change", event);
  expect(wrapper.find(".email").props().value).toEqual("shivam@gmail.com");
});

// test select tag for gender
fit("test gender select input", () => {
  wrapper
    .find(".gender")
    .simulate("change", { target: { value: GenderType.FEMALE } });
  console.log(wrapper.state());
  expect(wrapper.find(".gender").props().value).toBe(GenderType.FEMALE);
});

// test select tag for status
it("test status select input", () => {
  wrapper
    .find(".status")
    .simulate("change", { target: { value: StatusType.ACTIVE } });
  expect(wrapper.find(".status").props().value).toBe(StatusType.ACTIVE);
});

// Post
it("post test", () => {
  var MockAdapter = require("axios-mock-adapter");
  var mock = new MockAdapter(axios);

  const data = {
    name: "Harinarayan Abbott",
    email: "abbott_harinarayan@collins.info",
    gender: "male",
    status: "active",
  };

  const token =
    "3f30438c7b3212b121ae63e52bae216ca2bc11b700c8aa29cb0891d61cc96fca";

  mock.onPost("https://gorest.co.in/public/v2/users/").reply(200, data, {
    "Content-Type": "application/json",
    Authorization: `Bearer ${token}`,
  });

  let wrapper: any;
  act(() => {
    wrapper = shallow(<Adduser handleClose={jest.fn()} />);
  });

  wrapper.update();
  expect(toJson(wrapper)).toMatchSnapshot();
});

// submit button
it("should be able to submit the form", () => {
  const component = render(<Adduser handleClose={jest.fn()} />);
  const email = screen.getByPlaceholderText("Enter email id here");
  const button = screen.getAllByRole("Button");

  userEvent.type(email, "abhi@gmail.com");
  userEvent.click(button[0]);

  const user = screen.getByText("abhi@gmail.com");
  expect(user).toBeInTheDocument();
});
