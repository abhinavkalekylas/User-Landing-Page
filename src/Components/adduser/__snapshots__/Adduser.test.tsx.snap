// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post test 1`] = `
<div
  className="container"
>
  <Form
    className="form"
    onSubmit={[Function]}
    tag="form"
  >
    <FormGroup
      tag="div"
    >
      <div
        className="input input-group input-group-lg"
      >
        <label>
          Name :
        </label>
        <input
          aria-describedby="inputGroup-sizing-lg"
          aria-label="Sizing example input"
          autoComplete="off"
          className="username form-control my-2"
          name="name"
          onChange={[Function]}
          placeholder="Enter username here"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="input input-group input-group-lg"
      >
        <label>
          Email :
        </label>
        <input
          aria-describedby="inputGroup-sizing-lg"
          aria-label="Sizing example input"
          autoComplete="off"
          className="email form-control my-2"
          name="email"
          onChange={[Function]}
          placeholder="Enter email id here"
          required={true}
          type="email"
          value=""
        />
      </div>
      <div
        className="input input-group input-group-lg"
      >
        <label>
          Gender :
        </label>
        <select
          aria-label=".form-select-lg example"
          className="gender form-select form-select-lg mb-3"
        >
          <option
            value="male"
          >
            Male
          </option>
          <option
            value="female"
          >
            Female
          </option>
        </select>
      </div>
      <div
        className="input input-group input-group-lg"
      >
        <label>
          Status :
        </label>
        <select
          aria-label=".form-select-lg example"
          className="status form-select form-select-lg mb-3"
        >
          <option
            value="active"
          >
            Active
          </option>
          <option
            value="inactive"
          >
            Inactive
          </option>
        </select>
      </div>
      <Button
        className="btn_add"
        color="success"
        onClick={[MockFunction]}
        tag="button"
        type="submit"
      >
        Add User
      </Button>
    </FormGroup>
  </Form>
</div>
`;

exports[`renders correctly Adduser component 1`] = `
<BrowserRouter>
  <Router
    location={
      Object {
        "hash": "",
        "key": "default",
        "pathname": "/",
        "search": "",
        "state": null,
      }
    }
    navigationType="POP"
    navigator={
      Object {
        "action": "POP",
        "back": [Function],
        "block": [Function],
        "createHref": [Function],
        "forward": [Function],
        "go": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <AddUser
      handleClose={[MockFunction]}
    >
      <div
        className="container"
      >
        <Form
          className="form"
          onSubmit={[Function]}
          tag="form"
        >
          <form
            className="form"
            onSubmit={[Function]}
          >
            <FormGroup
              tag="div"
            >
              <div
                className="mb-3"
              >
                <div
                  className="input input-group input-group-lg"
                >
                  <label>
                    Name :
                  </label>
                  <input
                    aria-describedby="inputGroup-sizing-lg"
                    aria-label="Sizing example input"
                    autoComplete="off"
                    className="username form-control my-2"
                    name="name"
                    onChange={[Function]}
                    placeholder="Enter username here"
                    required={true}
                    type="text"
                    value=""
                  />
                </div>
                <div
                  className="input input-group input-group-lg"
                >
                  <label>
                    Email :
                  </label>
                  <input
                    aria-describedby="inputGroup-sizing-lg"
                    aria-label="Sizing example input"
                    autoComplete="off"
                    className="email form-control my-2"
                    name="email"
                    onChange={[Function]}
                    placeholder="Enter email id here"
                    required={true}
                    type="email"
                    value=""
                  />
                </div>
                <div
                  className="input input-group input-group-lg"
                >
                  <label>
                    Gender :
                  </label>
                  <select
                    aria-label=".form-select-lg example"
                    className="gender form-select form-select-lg mb-3"
                  >
                    <option
                      value="male"
                    >
                      Male
                    </option>
                    <option
                      value="female"
                    >
                      Female
                    </option>
                  </select>
                </div>
                <div
                  className="input input-group input-group-lg"
                >
                  <label>
                    Status :
                  </label>
                  <select
                    aria-label=".form-select-lg example"
                    className="status form-select form-select-lg mb-3"
                  >
                    <option
                      value="active"
                    >
                      Active
                    </option>
                    <option
                      value="inactive"
                    >
                      Inactive
                    </option>
                  </select>
                </div>
                <div
                  className="d-flex"
                >
                  <Button
                    className="btn_add"
                    color="success"
                    onClick={[MockFunction]}
                    tag="button"
                    type="submit"
                  >
                    <button
                      aria-label={null}
                      className="btn_add btn btn-success"
                      onClick={[Function]}
                      type="submit"
                    >
                      Add User
                    </button>
                  </Button>
                  <Button
                    className="reset_add"
                    color="secondary"
                    onClick={[Function]}
                    tag="button"
                  >
                    <button
                      aria-label={null}
                      className="reset_add btn btn-secondary"
                      onClick={[Function]}
                      type="button"
                    >
                      Reset
                    </button>
                  </Button>
                </div>
              </div>
            </FormGroup>
          </form>
        </Form>
      </div>
    </AddUser>
  </Router>
</BrowserRouter>
`;
